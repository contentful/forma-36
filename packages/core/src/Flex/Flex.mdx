[Flex](https://developer.mozilla.org/en-US/docs/Glossary/Flex) is a great tool when you wan't to line things up in one direction. For example boxes in the limited amout of space, that would wrap or not wrap. When it comes to Flex, browser calculate things only in one direction, each row at the time. When for example CSS Grid calculates always rows and columns at the same time. If you think CSS Grid component is what you are looking for go ahead and have a look on our [Grid component](../Grid/Grid.md)

## Table of contents

- [How to use Flex](#how-to-use-flex)
- [Code examples](#code-examples)

## How to use Flex

### Margins

Margins and padding for Flex component are based on our [spacing system](https://f36.contentful.com/foundation/spacing/).
Here is an overview of spacings that are availeble:

`spacing2Xs` | `spacingXs` | `spacingXs` | `spacingS` | `spacingM` | `spacingL` | `spacingXl` | `spacing2Xl` | `spacing3Xl` | `spacing4Xl`

Flex, as any other element, can get margin value, which will be added evenly to all of it's sides or more specific instructions like margin-top or margin-right. those 2 can not be added at the same time - they will overwrite itself. That's why, please remember to not use them together:

**e.g. WRONG**

```jsx
<Flex marginTop="spacingXs" margin="spacingXs">
  Element
</Flex>
```

**e.g. GOOD**

```jsx
<Flex marginTop="spacingXs" marginLeft="spacingXs">
  Element
</Flex>
```

## Code examples

```js
<Flex
  justifyContent="space-between"
  alignItems="center"
  marginBottom="spacingM"
>
  <Flex>Example item</Flex>
  <Flex>Example item</Flex>
  <Flex>Example item</Flex>
</Flex>
```

<br />

```jsx
<Flex
  justifyContent="space-between"
  alignItems="center"
  marginBottom="spacingM"
>
  <Flex>
    <Paragraph>User name</Paragraph>
  </Flex>
  <Flex>
    <Flex marginRight="spacingM">
      <TextLink href="http://google.com">User profile</TextLink>
    </Flex>
    <Flex>
      <Button>Add user to the team</Button>
    </Flex>
  </Flex>
</Flex>
```

<br />
