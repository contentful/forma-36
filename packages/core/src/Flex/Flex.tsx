import * as React from 'react';
import { css, cx } from 'emotion';
import {
  PolymorphicProps,
  PolymorphicComponent,
  ExpandProps,
} from '../Primitive/Primitive';
import { useBox } from '../Box';
import type { MarginProps, PaddingProps, CommonProps, Spacing } from '../types';
import type * as CSS from 'csstype';
import tokens from '@contentful/f36-tokens';

export interface FlexInternalProps
  extends CommonProps,
    MarginProps,
    PaddingProps {
  /**
   * Child nodes to be rendered in the component */
  children?: React.ReactNode;
  /**
   * Sets width: 100% */
  fullWidth?: boolean;
  /**
   * Sets height: 100% */
  fullHeight?: boolean;

  /**
   * Sets display: inline-flex */
  isInline?: boolean;

  /**
   * A shorthand property for flex-grow, flex-shrink, flex-basis
   */
  flex?: CSS.Property.Flex;
  /**
   * Defines the initial size of a flexbox item.
   * */
  flexBasis?: CSS.Property.FlexBasis;
  /**
   * Defines how much a flexbox item should shrink if there's not enough space available. */
  flexShrink?: CSS.Property.FlexShrink;
  /**
   * Defines if flexbox items appear on a single line or on multiple lines within a flexbox container. */
  flexWrap?: CSS.Property.FlexWrap;
  /**
   * Defines how flexbox items are ordered within a flexbox container. */
  flexDirection?: CSS.Property.FlexDirection;
  /**
   * Defines how much a flexbox item should grow if there's space available. */
  flexGrow?: CSS.Property.FlexGrow;
  /**
   * Defines a gap between flexbox items. */
  gap?: Spacing | string;
  /**
   * Defines how flexbox/grid items are aligned according to the main axis, within a flexbox/grid container. */
  justifyContent?: CSS.Property.JustifyContent;
  /**
   * Defines the default justify-self for all items of the box, giving them all a default way of justifying each box along the appropriate axis. */
  justifyItems?: CSS.Property.JustifyContent;
  /**
   * Sets the way a box is justified inside its alignment container along the appropriate axis. */
  justifySelf?: CSS.Property.JustifySelf;
  /**
   * Defines how flexbox items are aligned according to the cross axis, within a line of a flexbox container.
   */
  alignItems?: CSS.Property.AlignItems;
  /**
   * Works like align-items, but applies only to a single flexbox item, instead of all of them.
   */
  alignSelf?: CSS.Property.AlignItems;
  /**
   * Defines how each line is aligned within a flexbox/grid container.
   */
  alignContent?: CSS.Property.AlignContent;
  /**
   * Defines the order of a flexbox item
   */
  order?: CSS.Property.Order;
}

export type FlexProps<
  E extends React.ElementType = typeof FLEX_DEFAULT_TAG
> = PolymorphicProps<FlexInternalProps, E>;

const FLEX_DEFAULT_TAG = 'div';

function _Flex<E extends React.ElementType = typeof FLEX_DEFAULT_TAG>(
  {
    isInline,
    alignItems,
    alignSelf,
    alignContent,
    flex,
    flexBasis,
    flexShrink,
    flexDirection,
    flexGrow,
    flexWrap,
    fullHeight,
    fullWidth,
    gap,
    justifyContent,
    justifyItems,
    justifySelf,
    order,
    children,
    as,
    ...otherProps
  }: FlexProps<E>,
  ref: React.Ref<any>,
) {
  const { boxProps, Element } = useBox<React.ElementType>({
    ...otherProps,
    as: as || FLEX_DEFAULT_TAG,
  });
  return (
    <Element
      {...boxProps}
      className={cx(
        css({
          display: isInline ? 'inline-flex' : 'flex',
          width: fullWidth ? '100%' : undefined,
          height: fullHeight ? '100%' : undefined,
          flex,
          flexBasis,
          flexShrink,
          flexDirection,
          gap: tokens[gap] ?? gap,
          justifyContent,
          justifyItems,
          justifySelf,
          alignItems,
          alignSelf,
          alignContent,
          order,
          flexWrap,
          flexGrow,
        }),
        boxProps.className,
      )}
      ref={ref}
    >
      {children}
    </Element>
  );
}

export const Flex: PolymorphicComponent<
  ExpandProps<FlexInternalProps>,
  typeof FLEX_DEFAULT_TAG
> = React.forwardRef(_Flex);
