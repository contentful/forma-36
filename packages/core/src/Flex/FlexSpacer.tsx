import * as React from 'react';

import { Box, BoxProps } from '../Box';

export function FlexSpacer<E extends React.ElementType = 'div'>(
  props: BoxProps<E>,
) {
  return (
    <Box
      css={{
        justifySelf: 'center',
        alignSelf: 'center',
        flex: '1',
      }}
      {...props}
    />
  );
}
