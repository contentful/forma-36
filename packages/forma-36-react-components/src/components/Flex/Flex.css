.Flex {
  display: flex;
}

.Flex__inline {
  display: inline-flex;
}

.Flex__fullWidth {
  width: 100%;
}

.Flex__fullHeight {
  height: 100%;
}

.Flex__noShrink {
  flex-shrink: 0;
}
