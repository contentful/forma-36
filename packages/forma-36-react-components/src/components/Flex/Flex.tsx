import React from 'react';
import tokens from '@contentful/forma-36-tokens';
import cn from 'classnames';
import type * as CSS from 'csstype';

import styles from './Flex.css';

export type SpacingTypes =
  | 'none'
  | 'spacing2Xs'
  | 'spacingXs'
  | 'spacingS'
  | 'spacingM'
  | 'spacingL'
  | 'spacingXl'
  | 'spacing2Xl'
  | 'spacing3Xl'
  | 'spacing4Xl';

export interface FlexProps {
  /**
   * Class names to be appended to the className prop of the component */
  className?: string;
  /**
   * Child nodes to be rendered in the component */
  children?: React.ReactNode;
  /**
   * An ID used for testing purposes applied as a data attribute (data-test-id) */
  testId?: string;
  /**
   * Html tag used as container
   */
  htmlTag?: React.ElementType;
  /**
   * Sets width: 100% */
  fullWidth?: boolean;
  /**
   * Sets hegiht: 100% */
  fullHeight?: boolean;
  /**
   * Sets display:inline-flex */
  inlineFlex?: boolean;
  /**
   * Sets flex-shrink: 0 */
  noShrink?: boolean;
  /**
   * One of flex-wrap css values */
  flexWrap?: CSS.Property.FlexWrap;
  /**
   * One of flex-direction css values */
  flexDirection?: CSS.Property.FlexDirection;
  /**
   * One of flex-grow css values */
  flexGrow?: CSS.Property.FlexGrow;
  /**
   * One of justify-content css values */
  justifyContent?: CSS.Property.JustifyContent;
  /**
   * One of justify-content css values */
  justifyItems?: CSS.Property.JustifyContent;
  /**
   * One of justify-self css values */
  justifySelf?: CSS.Property.JustifySelf;
  /**
   * One of align-items css values */
  alignItems?: CSS.Property.AlignItems;
  /**
   * One of align-self css values */
  alignSelf?: CSS.Property.AlignItems;
  /**
   * sets margin to one of the corresponding spacing tokens, default is none */
  margin?: SpacingTypes;
  /**
   * sets margin-top to one of the corresponding spacing tokens, default is none */
  marginTop?: SpacingTypes;
  /**
   * sets margin-right to one of the corresponding spacing tokens, default is none */
  marginRight?: SpacingTypes;
  /**
   * sets margin-bottom to one of the corresponding spacing tokens, default is none */
  marginBottom?: SpacingTypes;
  /**
   * sets margin-left to one of the corresponding spacing tokens, default is none */
  marginLeft?: SpacingTypes;
  /**
   * sets padding to one of the corresponding spacing tokens, default is none */
  padding?: SpacingTypes;
  /**
   * sets padding-top to one of the corresponding spacing tokens, default is none */
  paddingTop?: SpacingTypes;
  /**
   * sets padding-right to one of the corresponding spacing tokens, default is none */
  paddingRight?: SpacingTypes;
  /**
   * sets padding-bottom to one of the corresponding spacing tokens, default is none */
  paddingBottom?: SpacingTypes;
  /**
   * sets padding-left to one of the corresponding spacing tokens, default is none */
  paddingLeft?: SpacingTypes;
  /**
   * style prop, for inline styles */
  style?: React.CSSProperties;
}

const handleSpacing = (value: SpacingTypes) =>
  value === 'none' ? 0 : tokens[value];

const defaultProps = {
  testId: 'cf-ui-flex',
};

export const Flex = (props: FlexProps) => {
  const {
    htmlTag: Container = 'div',
    fullWidth,
    fullHeight,
    noShrink,
    flexWrap,
    flexDirection,
    flexGrow,
    inlineFlex,
    justifyContent,
    justifyItems,
    justifySelf,
    alignItems,
    alignSelf,
    margin,
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    padding,
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft,
    className,
    children,
    testId,
    style,
    ...otherProps
  } = props;

  const fullMargins = { margin: margin && handleSpacing(margin) };
  const sidesMargins = {
    marginTop: marginTop && handleSpacing(marginTop),
    marginRight: marginRight && handleSpacing(marginRight),
    marginBottom: marginBottom && handleSpacing(marginBottom),
    marginLeft: marginLeft && handleSpacing(marginLeft),
  };

  const fullPaddings = { padding: padding && handleSpacing(padding) };
  const sidesPaddings = {
    paddingTop: paddingTop && handleSpacing(paddingTop),
    paddingRight: paddingRight && handleSpacing(paddingRight),
    paddingBottom: paddingBottom && handleSpacing(paddingBottom),
    paddingLeft: paddingLeft && handleSpacing(paddingLeft),
  };
  const marginResult = margin ? fullMargins : sidesMargins;
  const paddingsResult = padding ? fullPaddings : sidesPaddings;

  const inlineStyle = {
    flexDirection,
    justifyContent,
    justifyItems,
    justifySelf,
    alignItems,
    alignSelf,
    flexWrap,
    flexGrow,
  };

  const classNames = cn(styles.Flex, className, {
    [styles.Flex__inline]: inlineFlex,
    [styles.Flex__fullWidth]: fullWidth,
    [styles.Flex__fullHight]: fullHeight,
    [styles.Flex__noShrink]: noShrink,
  });

  return (
    <Container
      {...otherProps}
      style={{ ...inlineStyle, ...marginResult, ...paddingsResult, ...style }}
      className={classNames}
      data-test-id={testId}
    >
      {children}
    </Container>
  );
};
Flex.defaultProps = defaultProps;

export default Flex;
