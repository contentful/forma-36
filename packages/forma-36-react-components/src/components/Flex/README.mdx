# Flex


[Flex](https://developer.mozilla.org/en-US/docs/Glossary/Flex) is a great tool when you wan't to line things up in one direction. For example boxes in the limited amout of space, that would wrap or not wrap. When it comes to Flex, browser calculate things only in one direction, each row at the time. When for example CSS Grid calculates always rows and columns at the same time. If you think CSS Grid component is what you are looking for go ahead and have a look on our [Grid component](../Grid/Grid.md)

## Usage

```js
import { Flex } from '@contenful/forma-36-react-components';
// if component is still in alpha stage
// import { Flex } from '@contenful/forma-36-react-components/dist/alpha';

<Flex justifyContent="space-between" alignItems="center" marginBottom="spacingM">
  <Flex>Example item</Flex>
  <Flex>Example item</Flex>
  <Flex>Example item</Flex>
</Flex>

```
<br />

```jsx
import { Flex, Button, TextLink, Paragraph } from '@contenful/forma-36-react-components';
// if component is still in alpha stage
// import { Flex } from '@contenful/forma-36-react-components/dist/alpha';

<Flex justifyContent="space-between" alignItems="center" marginBottom="spacingM">
  <Flex><Paragraph>User name</Paragraph></Flex>
  <Flex>
    <Flex marginRight="spacingM">
      <TextLink href="http://google.com">User profile</TextLink>
    </Flex>
    <Flex><Button>Add user to the team</Button></Flex>
  </Flex>
</Flex>
```
<br />

### Margins & paddings

Margins and padding for Flex component are based on our [spacing system](https://f36.contentful.com/foundation/spacing/).
Here is an overview of spacings that are availeble:

`spacing2Xs` | `spacingXs` | `spacingXs` | `spacingS` | `spacingM` | `spacingL` | `spacingXl` | `spacing2Xl` | `spacing3Xl` | `spacing4Xl`

Flex, as any other element, can get margin value, which will be added evenly to all of it's sides or more specific instructions like margin-top or margin-right. those 2 can not be added at the same time - they will overwrite itself. That's why, please remember to not use them together:

**e.g. WRONG**

```jsx
import { Flex } from '@contenful/forma-36-react-components';
// if component is still in alpha stage
// import { Flex } from '@contenful/forma-36-react-components/dist/alpha';

  <Flex marginTop='spacingXs' margin="spacingXs">Element</Flex>
```
**e.g. GOOD**

```jsx
import { Flex } from '@contenful/forma-36-react-components';
// if component is still in alpha stage
// import { Flex } from '@contenful/forma-36-react-components/dist/alpha';

  <Flex marginTop='spacingXs' marginLeft="spacingXs">Element</Flex>
```

Component props overview:

|Prop|Description|
|-|-|
|`htmlTag`| string, one of `a` `article`  `aside`  `div`  `footer`  `form`  `h1`  `h2`  `h3`  `h4`  `h5`  `h6`  `header`  `input`  `li`  `main`  `nav`  `ol`  `p`  `pre`  `section`  `span`  `textarea`  `ul` |
|`fullWidth`| boolean, sets width to 100% |
|`fullHeight`| boolean, sets height to 100% |
|`inlineFlex`| boolean, sets display to `inline-flex` |
|`noShrink`| boolean, sets flex-shrink to 0  - default is 1|
|`flexWrap`| string, one of `flex-wrap` css values |
|`flexDirection`| string, one of `flex-direction` css values |
|`justifyItems`| string, one of `justify-items` css values |
|`justifyContent`| string, one of `justify-content` css values |
|`justifySelf`| string, one of `justify-self` css values |
|`alignItems`| string, one of `align-items` css values |
|`alignSelf`| string, one of `align-self` css values |
|`margin`| string, sets `margin` to one of the spacing values |
|`marginTop`| string, sets `margin-top` to one of the spacing values |
|`marginRight`| string, sets `margin-right` to one of the spacing values |
|`marginBottom`| string, sets `margin-bottom` to one of the spacing values |
|`marginLeft`| string, sets `margin-left` to one of the spacing values |
|`padding`| string, sets `padding` to one of the spacing values |
|`paddingTop`| string, sets `padding-top` to one of the spacing values |
|`paddingRight`| string, sets `padding-right` to one of the spacing values |
|`paddingBottom`| string, sets `padding-bottom` to one of the spacing values |
|`paddingLeft`| string, sets `padding-left` to one of the spacing values |
