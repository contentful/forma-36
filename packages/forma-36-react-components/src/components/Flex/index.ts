export * from './Flex';
export { default } from './Flex';
