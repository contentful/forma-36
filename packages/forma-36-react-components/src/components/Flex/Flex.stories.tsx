import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, number, select } from '@storybook/addon-knobs';
import tokens from '@contentful/forma-36-tokens';
import Flex from './Flex';
import notes from './README.md';

const flexKnobsId = 'Flex Props';
const spacingValues = [
  'none',
  'spacing2Xs',
  'spacingXs',
  'spacingS',
  'spacingM',
  'spacingL',
  'spacingXl',
  'spacing2Xl',
  'spacing3Xl',
  'spacing4Xl',
];
const flexDirectionValues = {
  'column': 'column',
  'column-reverse': 'column-reverse',
  'row': 'row',
  'row-reverse': 'row-reverse',
};

const flexJustifyValues = {
  'center': 'center',
  'start':'start',
  'end':'end',
  'flex-start':'flex-start',
  'flex-end':'flex-end',
  'space-between':'space-between',
  'space-around':'space-around',
  'space-evenly':'space-evenly',
  'stretch':'stretch',
}

const flexAlignValues = {
  'center': 'center',
  'start':'start',
  'end':'end',
  'flex-start':'flex-start',
  'flex-end':'flex-end',
  'stretch':'stretch',
}

const flexWrapValues = {
  "nowrap": 'nowrap', 
  "wrap": 'wrap',
  "wrap-reverse": 'wrap-reverse'
};

const styles = {
  demoBox: {
    backgroundColor: tokens.colorContrastLight,
    width: '150px',
    height: '80px',
    color: tokens.colorWhite
  },
};

const DemoBox = ({
  times,
}: {
  times?: number,
}) => {
  if (times) {
    let result = [];
    for (let i = 0; i < times; i++) {
    result.push(
      <Flex 
        style={styles.demoBox}
        justifyContent="center"
        alignItems="center"
        padding="spacingM"
        margin="spacingXs"
      >
          Example element {i}
      </Flex>
    );
    }
    return <>{result}</>;
  }
  return <Flex style={styles.demoBox}>Example element</Flex>;
};

storiesOf('(alpha)|Flex', module)
  .addParameters({
    propTypes: Flex['__docgenInfo'],
  })
  .add(
    'default',
    () => {
      const knobs = {
        exampleBoxes: number('Example Items', 4),
        flexDirection: select('Flex direction', flexDirectionValues, 'row'),
        justifyContent: select('Justify content', flexJustifyValues, 'center'),
        alignItems: select('Align items', flexAlignValues, 'center'),
        flexWrap: select('Flex wrap', flexWrapValues, 'center'),
      };

      return (
        <Flex
          margin="spacing4Xl"
          flexDirection={knobs.flexDirection}
          justifyContent={knobs.justifyContent}
          alignItems={knobs.alignItems}
          flexWrap={knobs.flexWrap}
          fullWidth={boolean('Full width', false)}
          fullHeight={boolean('Full height', false)}
          inlineFlex={boolean('Inline flex', false)}
          noShrink={boolean('No shrink', false)}
          htmlTag="article"
        >
          <DemoBox times={knobs.exampleBoxes} />
        </Flex>
      );
    },
    { notes },
  )
  .add(
    'Flex margins',
    () => {
      const knobs = {
        exampleBoxes: number('Example Items', 4),
        margin: select('Margin', spacingValues, 'spacingXs', flexKnobsId),
        marginTop: select('Margin top', spacingValues, 'none', flexKnobsId),
        marginRight: select('Margin right', spacingValues, 'none', flexKnobsId),
        marginBottom: select('Margin bottom', spacingValues, 'none', flexKnobsId),
        marginLeft: select('Margin left', spacingValues, 'none', flexKnobsId),
      };

      return (
        <Flex
          htmlTag="div"
        >
          <Flex
            margin={knobs.margin !== 'none' ? knobs.margin : null}
            marginTop={knobs.marginTop !== 'none' ? knobs.marginTop : null}
            marginRight={knobs.marginRight !== 'none' ? knobs.marginRight : null}
            marginBottom={knobs.marginBottom !== 'none' ? knobs.marginBottom : null}
            marginLeft={knobs.marginLeft !== 'none' ? knobs.marginLeft : null}
            style={styles.demoBox}
            justifyContent="center"
            alignItems="center"
          >Example element 1</Flex>
          <Flex
            margin={knobs.margin !== 'none' ? knobs.margin : null}
            marginTop={knobs.marginTop !== 'none' ? knobs.marginTop : null}
            marginRight={knobs.marginRight !== 'none' ? knobs.marginRight : null}
            marginBottom={knobs.marginBottom !== 'none' ? knobs.marginBottom : null}
            marginLeft={knobs.marginLeft !== 'none' ? knobs.marginLeft : null}
            style={styles.demoBox}
            justifyContent="center"
            alignItems="center"
          >Example element 2</Flex>
        </Flex>
      );
    },
    { notes },
  )
  .add(
    'Flex paddings',
    () => {
      const knobs = {
        exampleBoxes: number('Example Items', 4),
        padding: select('Padding', spacingValues, 'spacingXs', flexKnobsId),
        paddingTop: select('Padding top', spacingValues, 'none', flexKnobsId),
        paddingRight: select('Padding right', spacingValues, 'none', flexKnobsId),
        paddingBottom: select('Padding bottom', spacingValues, 'none', flexKnobsId),
        paddingLeft: select('Padding left', spacingValues, 'none', flexKnobsId),
      };

      return (
        <Flex
          htmlTag="div"
        >
          <Flex
            padding={knobs.padding !== 'none' ? knobs.padding : null}
            paddingTop={knobs.paddingTop !== 'none' ? knobs.paddingTop : null}
            paddingRight={knobs.paddingRight !== 'none' ? knobs.paddingRight : null}
            paddingBottom={knobs.paddingBottom !== 'none' ? knobs.paddingBottom : null}
            paddingLeft={knobs.paddingLeft !== 'none' ? knobs.paddingLeft : null}
            style={styles.demoBox}
            justifyContent="center"
            alignItems="center"
            marginRight="spacingXs"
          >Example element 1</Flex>
          <Flex
            padding={knobs.padding !== 'none' ? knobs.padding : null}
            paddingTop={knobs.paddingTop !== 'none' ? knobs.paddingTop : null}
            paddingRight={knobs.paddingRight !== 'none' ? knobs.paddingRight : null}
            paddingBottom={knobs.paddingBottom !== 'none' ? knobs.paddingBottom : null}
            paddingLeft={knobs.paddingLeft !== 'none' ? knobs.paddingLeft : null}
            style={styles.demoBox}
            justifyContent="center"
            alignItems="center"
          >Example element 2</Flex>
        </Flex>
      );
    },
    { notes },
  );
