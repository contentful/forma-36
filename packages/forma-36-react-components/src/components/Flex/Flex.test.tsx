import React from 'react';
import { shallow, mount } from 'enzyme';
import axe from '../../utils/axeHelper';
import Flex from './Flex';

it('renders the component', () => {
  const output = shallow(<Flex>Flex</Flex>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(<Flex className="my-extra-class">Flex</Flex>);

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<Flex>Flex</Flex>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});

describe('should have correct styles', () => {
  it('justifyContent should be center by default', () => {
    const output = shallow(<Flex justifyContent="center">Flex</Flex>);
    const outputStyles = output.get(0).props.style;
    expect(outputStyles).toMatchObject({ justifyContent: 'center' });
  });

  it('alignItems should be center by default', () => {
    const output = shallow(<Flex alignItems="center">Flex</Flex>);
    const outputStyles = output.get(0).props.style;
    expect(outputStyles).toMatchObject({ alignItems: 'center' });
  });

  it('should render correct margin value', () => {
    const output = shallow(<Flex marginRight="spacingXs">Flex</Flex>);
    const outputStyles = output.get(0).props.style;
    expect(outputStyles).toMatchObject({
      marginRight: '0.5rem',
    });
  });

  it('should render correct padding value', () => {
    const output = shallow(<Flex paddingRight="spacingXs">Flex</Flex>);
    const outputStyles = output.get(0).props.style;
    expect(outputStyles).toMatchObject({
      paddingRight: '0.5rem',
    });
  });
});
