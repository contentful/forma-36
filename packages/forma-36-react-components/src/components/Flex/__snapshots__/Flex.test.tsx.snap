// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  className="Flex"
  data-test-id="cf-ui-flex"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": undefined,
      "flexDirection": "row",
      "flexWrap": undefined,
      "justifyContent": "center",
      "justifyItems": undefined,
      "justifySelf": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
    }
  }
>
  Flex
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  className="Flex my-extra-class"
  data-test-id="cf-ui-flex"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": undefined,
      "flexDirection": "row",
      "flexWrap": undefined,
      "justifyContent": "center",
      "justifyItems": undefined,
      "justifySelf": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
    }
  }
>
  Flex
</div>
`;
